### PR TITLE
DAOS-1627 raft: Membership Changes

### PR DESCRIPTION
**Pull request by okulkarni**
1. Fixed errors arising from repeated membership changes.

2. Changes to allow operating with an empty configuration.

3. Removed a redundant assignment and miscellaneous cleanup.

4. Added / updated test cases.

Signed-off-by: Omkar Kulkarni <omkar.kulkarni@intel.com>

### CODE REVIEW

File: src/raft_server.c [hunk 1]
```diff
@@ -1084,6 +1097,7 @@ static raft_node_t* raft_add_non_voting_node_internal(raft_server_t* me_, raft_e
     if (!node)
         return NULL;
 
+    raft_node_set_addition_committed(node, 0);
```
Comment (liw) [inline src/raft_server.c]: Looks like it would be more intuitive to keep raft_node_new unchanged...

File: src/raft_server.c [hunk 2]
```diff
@@ -933,12 +948,10 @@ int raft_apply_entry(raft_server_t* me_)
...
             break;
         case RAFT_LOGTYPE_DEMOTE_NODE:
-            if (node)
-                raft_node_set_voting_committed(node, 0);
+            raft_node_set_voting_committed(node, 0);
             break;
         case RAFT_LOGTYPE_REMOVE_NODE:
-            if (node)
-                raft_remove_node(me_, node);
+            raft_node_set_addition_committed(node, 0);
```
Comment (liw) [inline src/raft_server.c]: What cases need this and those changes in raft_apply_all? The membership change algorithm allows at most one uncommitted membership change entry, so we can simply remove the node here, no? (I must have missed your considerations.)

Comment (okulkarni) [reply src/raft_server.c]: When we're making live changes - yes!

But if we add a new node and it receives a series of membership change entries as part of "appendentries" from the leader it has to commit and apply all of them locally. The best way I figured out was to keep track of additions and removals and making changing the actual node list at the end (deleting all nodes which have been marked for removal, i.e. addition_committed == false). 

This is a little tricky to understand if we don't see RDB/raft in action. The problem is that if we have log entries affecting nodes ranks 'x', 'y' and 'z' -- we need those nodes to be present in the node list even before those log entries have been applied. 

Let me know if you'd like to discuss this further.

File: src/raft_server.c (hunk 1 above)
Comment (okulkarni) [reply src/raft_server.c]: There's a difference: 

raft_add_[non_voting]_node() is called in response to a log entry. We don't want the addition to be "finalized" until that log entry has been applied.

But raft_node_new() is a private call -- for example to create an initial list of replicas.

File: src/raft_server.c [hunk 3]
```diff
@@ -917,8 +928,12 @@ int raft_apply_entry(raft_server_t* me_)
     if (!raft_entry_is_cfg_change(ety))
         return 0;
 
+    if(!me->cb.log_get_node_id)
+        return 0;
+
```
[the remark this answers is not in the document]
Comment (liw) [reply src/raft_server.c]: I'm afraid we can't set commitIndex to lastIndex in rdb_raft_start. We need to find out and address "the issues". :(

File: src/raft_server.c (hunk 2 above)
Comment (liw) [reply src/raft_server.c]: I think I've gained a feeling of the issue. The membership change handling in raft_apply_entry appears to cause both the issue you described here and the issue in rdb_raft_start. By the time we apply a membership change entry i in raft_apply_entry, the node in question may have already been changed by a latter entry j when j is offered.

**Pull request closed without merging**